### PR TITLE
Adding CAS functionality to the consul-kv library

### PR DESCRIPTION
**Pull request by jakobdalsgaard**
CAS for Consul is the ability to guard updates in the K/V against updates from multiple clients. The implementation requires for the client to pull out a version number with each data item, and supply that version number with each update or delete.
If some else has modified that piece of data then version number has changed, and update is denied.

### CODE REVIEW

File: consul_kv/api.py [hunk 1]
```diff
@@ -76,33 +90,36 @@ def put_kv_txn(mapping, endpoint=DEFAULT_TXN_ENDPOINT, timeout=socket._GLOBAL_DE
...
-def delete_kv(k=None, recurse=False, endpoint=DEFAULT_KV_ENDPOINT, timeout=socket._GLOBAL_DEFAULT_TIMEOUT):
+def get_kv_builder(postprocessor=None):
+    def get_kv_raw(k=None, recurse=False, endpoint=DEFAULT_KV_ENDPOINT, timeout=socket._GLOBAL_DEFAULT_TIMEOUT):
+        url = join(endpoint, k) if k else endpoint
+        url = "%s?recurse"  % url if recurse else url
+        req = request.Request(
+                url=url,
+                method='GET'
+        )
+        print(" fetching %s %s" % (k, url))
```
Comment (vdloo) [inline consul_kv/api.py]: I think you left some debug statements here and above, can you remove those?

File: consul_kv/api.py [hunk 2]
```diff
@@ -76,33 +90,36 @@ def put_kv_txn(mapping, endpoint=DEFAULT_TXN_ENDPOINT, timeout=socket._GLOBAL_DE
...
-        if r['Value'] else None for r in result if r['Key']
-    }
-    return mapping
-
-
-def delete_kv(k=None, recurse=False, endpoint=DEFAULT_KV_ENDPOINT, timeout=socket._GLOBAL_DEFAULT_TIMEOUT):
+def get_kv_builder(postprocessor=None):
+    def get_kv_raw(k=None, recurse=False, endpoint=DEFAULT_KV_ENDPOINT, timeout=socket._GLOBAL_DEFAULT_TIMEOUT):
+        url = join(endpoint, k) if k else endpoint
+        url = "%s?recurse"  % url if recurse else url
```
Comment (vdloo) [inline consul_kv/api.py]: I would prefer it if you used the new style string formatting with `{}` and `.format()`

Comment (jakobdalsgaard) [inline consul_kv/api.py]: OK :-)

**Comment by jakobdalsgaard** [comment]
Now I also found out about autopep8 :-)

**Comment by vdloo** [comment]
yeah there's some pep8 stuff, I should set up travis for this repo:
```
./runtests.sh -1
Running consul-kv unit tests
...............................................................
----------------------------------------------------------------------
Ran 63 tests in 0.186s

OK
Checking PEP8
--- original/consul_kv/__init__.py
+++ fixed/consul_kv/__init__.py
@@ -58,15 +58,15 @@
 
     def get_cas(self, k=None, recurse=False):
         return get_kv_cas(
-                k=k, recurse=recurse, endpoint=self.kv_endpoint,
-                timeout=self.timeout
-                )
+            k=k, recurse=recurse, endpoint=self.kv_endpoint,
+            timeout=self.timeout
+        )
 
     def get_meta(self, k=None, recurse=False):
         return get_kv_meta(
-                k=k, recurse=recurse, endpoint=self.kv_endpoint,
-                timeout=self.timeout
-                )
+            k=k, recurse=recurse, endpoint=self.kv_endpoint,
+            timeout=self.timeout
+        )
 
     def get(self, k=None, recurse=False):
         """
--- original/consul_kv/api.py
+++ fixed/consul_kv/api.py
@@ -64,6 +64,7 @@
         } for k, v in mapping.items()
     ]
 
+
 def put_kv_txn(mapping, endpoint=DEFAULT_TXN_ENDPOINT, timeout=socket._GLOBAL_DEFAULT_TIMEOUT):
     """
     Update multiple keys inside a single, atomic transaction.
@@ -93,10 +94,10 @@
 def get_kv_builder(postprocessor=None):
     def get_kv_raw(k=None, recurse=False, endpoint=DEFAULT_KV_ENDPOINT, timeout=socket._GLOBAL_DEFAULT_TIMEOUT):
         url = join(endpoint, k) if k else endpoint
-        url = "%s?recurse"  % url if recurse else url
+        url = "%s?recurse" % url if recurse else url
         req = request.Request(
-                url=url,
-                method='GET'
+            url=url,
+            method='GET'
         )
         print(" fetching %s %s" % (k, url))
         with request.urlopen(req, timeout=timeout) as r:
@@ -107,18 +108,18 @@
 get_kv_meta = get_kv_builder(lambda x: x)
 
 get_kv_cas = get_kv_builder(lambda x: {
-            r['Key']: (b64decode(r['Value']).decode('utf-8'), r['ModifyIndex'])
-            if r['Value'] else None for r in x if r['Key']
-          })
+    r['Key']: (b64decode(r['Value']).decode('utf-8'), r['ModifyIndex'])
+    if r['Value'] else None for r in x if r['Key']
+})
 
 get_kv = get_kv_builder(lambda x: {
-            # values are stored base64 encoded in consul, they
-            # are decoded before returned by this function.
-            r['Key']: b64decode(r['Value']).decode('utf-8')
-            if r['Value'] else None for r in x if r['Key']
-          })
+    # values are stored base64 encoded in consul, they
+    # are decoded before returned by this function.
+    r['Key']: b64decode(r['Value']).decode('utf-8')
+    if r['Value'] else None for r in x if r['Key']
+})
 
-      
+
 def delete_kv(k=None, cas=None, recurse=False, endpoint=DEFAULT_KV_ENDPOINT, timeout=socket._GLOBAL_DEFAULT_TIMEOUT):
     """
     Delete a key from the distributed key value store
@@ -137,7 +138,6 @@
     if len(params) > 0:
         url = "%s?%s" % (url, parse.urlencode(params))
 
-
     req = request.Request(
         url="%s?recurse" % url if recurse else url,
         method='DELETE'

```

**Comment by jakobdalsgaard** [comment]
My pep8 run on the code runs smoothly now

**Pull request merged**